fix(frontend): detect a connected user in a safe

compojoom opened the pull request:
# Summary
If the user navigates to the CoW app in safe wallet and the user is not connected to the Safe, on connection the CoW site doesn’t detect this and ask the user to refresh the page. By polling the safeInfo we can detect that the connected safe changed state.

![grafik](https://github.com/cowprotocol/cowswap/assets/693770/8e08910e-5759-4ae5-8394-29c49a14ea90)

It actually should no longer be necessary to refresh.

# To Test
1. Navigate to a safe (where not connected)
2. Connect when on the widget/app page
3. After few sec. CoW should change the "connect signer" text to "select token"

anxolin commented on libs/wallet/src/web3-react/hooks/useSafeAppsSdkInfo.ts:
```diff
@@ -9,14 +9,23 @@ export type GnosisSafeSdkInfo = SafeInfo
+    }, intervalTimeout)
+
+    return () => clearInterval(intervalId)
+  }, [safeAppsSdk, intervalTimeout])
```
This is a great idea.

One concern I have is that this logic is done in a very general hook. Hooks when used duplicate the logic, meaning that this polling will be done on every instance of the hook. Probably not what you want to do

Look here:
![image](https://github.com/cowprotocol/cowswap/assets/2352112/75c66a15-b2c1-4c68-9e0d-2418d07ac1c5)

Maybe is better to do this once for all the app. I could think `WalletUpdater` is a better place to do it. 

One approach could be for `useSafeAppsSdkInfo()` to return a function to refresh the gnosis safe info on demand. WDYT?

anxolin commented on libs/wallet/src/web3-react/hooks/useSafeAppsSdkInfo.ts:
```diff
@@ -9,14 +9,23 @@ export type GnosisSafeSdkInfo = SafeInfo
+        safeAppsSdk.safe.getInfo().then((safeInfo) => {
+          setGnosisSafeInfo(safeInfo)
+          // if the user is connected, we can check less frequently
+          setIntervalTimeout(safeInfo.isReadOnly ? 10000 : 2000)
```
could you use consts for the interval delays?

anxolin reviewed (commented):
Awesome you took the initiative!

anxolin commented on libs/wallet/src/web3-react/hooks/useSafeAppsSdkInfo.ts:
```diff
@@ -1,22 +1,24 @@
 export function useSafeAppsSdkInfo(): GnosisSafeSdkInfo | null {
-  const [gnosisSafeInfo, setGnosisSafeInfo] = useState<GnosisSafeSdkInfo | null>(null)
-  const safeAppsSdk = useSafeAppsSdk()
+  const gnosisSafeInfo = useAtom(gnosisSafeInfoAtom)[0]
```
If you just need to READ the atom, you should use `useAtomValue`

anxolin replied:
This hooks was exposing the SDK information extracted form the connector. 
Now its exposing the global state.

I think this helped with the re-renders, but IMO is not correct. 

A first small issue is that being called SafeAppsSdkInfo should return the info coming from the SDK, but most of its info comes from the API. This is conceptually misleading. 

What is right, is that we should expose this `gnosisSafeInfo` using a hook, AND whomever is using `useSafeAppsSdkInfo` should use that one instead. 

Some other small thing, is, the implementation to export the global info should be simpler:
something like:
```
export function useSafeInfo() {
    return useAtomValue(safeInfoAtom)
}
```

then hooks like `useIsSafeWallet()` should use `useSafeInfo()` and not `useSafeAppsSdkInfo()`

anxolin commented on libs/wallet/src/web3-react/hooks/useSafeAppsSdkInfo.ts:
```diff
@@ -1,22 +1,24 @@
+    chainId: gnosisSafeInfo.chainId,
+    threshold: gnosisSafeInfo.threshold,
+    owners: gnosisSafeInfo.owners,
+    isReadOnly: !!gnosisSafeInfo.isReadOnly,
```
nit: destructure

anxolin replied:
my comment is not relevant, cause I think we should implement this differently (explained above my thoughts)

anxolin commented on libs/wallet/src/web3-react/updater.ts:
```diff
@@ -17,8 +17,9 @@ import { gnosisSafeInfoAtom, walletDetailsAtom, walletInfoAtom } from '../api/st
+import { useSafeAppsSdk } from './hooks/useSafeAppsSdk'
 
-const SAFE_INFO_UPDATE_INTERVAL = ms`30s`
+const SAFE_INFO_UPDATE_INTERVAL = ms`5s`
```
This for testing is good, but we will need to reduce it, so Uxio don't get upset

Uxio0 replied:
Well thought

compojoom replied:
love is in the air :D!

anxolin commented on libs/wallet/src/web3-react/updater.ts:
```diff
@@ -65,17 +66,27 @@ function _useSafeInfo(walletInfo: WalletInfo): GnosisSafeInfo | undefined {
+                ..._safeInfo,
+              })
+            } else {
+              setSafeInfo({
```
this admits a refactor, as the only thing change is isReadOny

anxolin commented on libs/wallet/src/web3-react/updater.ts:
```diff
@@ -65,17 +66,27 @@ function _useSafeInfo(walletInfo: WalletInfo): GnosisSafeInfo | undefined {
-            })
+          .then(async (_safeInfo) => {
+            if (safeAppsSdk) {
+              const appsSdkSafeInfo = await safeAppsSdk.safe.getInfo()
```
I like this, when we update, we update all

anxolin replied:
Probably with this, we don't even need to have a `useSafeAppsSdkInfo()` at all

compojoom commented:
@anxolin please check again. I think that now it actually makes sense! 
I've removed the useSafeAppsSdkInfo! And changed the locations where it is used to `useGnosisSafeInfo`. At the same time the updater makes sure that we have the current isReadOnly state and it is polling every 5s. It doesn't seem that you need anything else from the useSafeAppsSdkInfo.

alfetopito commented on libs/wallet/src/web3-react/hooks/useWalletMetadata.ts:
```diff
@@ -125,10 +124,9 @@ export function useIsSafeApp(): boolean {
   const hasSafeInfo = !!useGnosisSafeInfo()
 
-  return !!safeAddress || hasSafeInfo
+  return hasSafeInfo
```
Nitpick:

```suggestion
  return !!useGnosisSafeInfo()
```

alfetopito commented on libs/wallet/src/web3-react/updater.ts:
```diff
@@ -65,18 +65,22 @@ function _useSafeInfo(walletInfo: WalletInfo): GnosisSafeInfo | undefined {
-  const { isReadOnly } = useSafeAppsSdkInfo() || {}
+  const safeAppsSdk = useSafeAppsSdk()
+
+  const SAFE_INFO_UPDATE_INTERVAL = ms`30s`
```
This constant is being re-instantiated from the outer scope

alfetopito commented on libs/wallet/src/web3-react/updater.ts:
```diff
@@ -86,19 +90,39 @@ function _useSafeInfo(walletInfo: WalletInfo): GnosisSafeInfo | undefined {
     }
 
-    const interval = setInterval(update, SAFE_INFO_UPDATE_INTERVAL)
+    const updateSafeAppsSdkInfo = async () => {
```
Since this should only update `isReadOnly`, I suggest to make the name more explicit:

```suggestion
    const updateIsReadOnlyFromSafeAppsSdkInfo = async () => {
```

alfetopito commented on libs/wallet/src/web3-react/updater.ts:
```diff
@@ -86,19 +90,39 @@ function _useSafeInfo(walletInfo: WalletInfo): GnosisSafeInfo | undefined {
+        const appsSdkSafeInfo = await safeAppsSdk.safe.getInfo()
+        setSafeInfo((prevSafeInfo) => {
+          // only update the isReadyOnly flag if we have safe info returned from the api kit
+          if (!prevSafeInfo) return
+          return {
+            ...prevSafeInfo,
+            isReadOnly: appsSdkSafeInfo.isReadOnly,
+          }
```
Small optimization idea: only update if values are different:

```suggestion
          if (!prevSafeInfo || prevSafeInfo.isReadOnly === appsSdkSafeInfo.isReadOnly) return

          return {
            ...prevSafeInfo,
            isReadOnly: appsSdkSafeInfo.isReadOnly,
          }
```

compojoom replied:
I thought about this, but then decided it's not worth it. React shouldn't trigger a re-render if the state is not changing...

alfetopito commented on libs/wallet/src/web3-react/updater.ts:
```diff
@@ -86,19 +90,39 @@ function _useSafeInfo(walletInfo: WalletInfo): GnosisSafeInfo | undefined {
 
-    return () => clearInterval(interval)
-  }, [setSafeInfo, chainId, account, provider, isReadOnly])
+    updateSafeInfo()
+    updateSafeAppsSdkInfo()
```
Is it possible for them to be out of sync in regards to `isReadOnly`?
For example, what would happen if the sdkInfo returns `false`, then another update from `getSafeInfo` runs and it's marked as `true`?

compojoom replied:
Yes, I thought about this. Since both updates use the functional update pattern, then it is guaranteed that at the time of running those functions would have the most up to date previous state. 

This doesn't really help with the potential race condition, but I think that it should be fine:
- updateSafeInfo doesn't touch the isReadOnly flag. it sets the previous state, then overrides it with the values from getSafeInfo, note that getSafeInfo doesn't return an isReadOnly flag, so in reality the only function that can update this state is updateIsReadOnlyFromSafeAppsSdkInfo. Makes sense?

alfetopito replied:
Ah ok. Then it's fine.

alfetopito commented on libs/wallet/src/web3-react/updater.ts:
```diff
@@ -65,18 +65,22 @@ function _useSafeInfo(walletInfo: WalletInfo): GnosisSafeInfo | undefined {
+  const safeAppsSdk = useSafeAppsSdk()
+
+  const SAFE_INFO_UPDATE_INTERVAL = ms`30s`
+  const SAFE_APPS_SDK_INFO_UPDATE_INTERVAL = ms`5s` // Adjust this value as needed
```
If this is meant as a constant, please declare it in the global scope.

alfetopito reviewed (commented):
Thanks again!

I left a few comments and suggestions

alfetopito reviewed (approved):
I tested and it's working great!

One last point I noticed that needs update:

![image](https://github.com/cowprotocol/cowswap/assets/43217/d3592353-7eb6-4a44-84c7-377b170ede03)

Just remove the `and refresh the page.` bit.

compojoom commented:
@alfetopito - thanks! Updated. Do you want me to squash all commits? Or you are going to do this anyway when you merge?

alfetopito commented:
> @alfetopito - thanks! Updated. Do you want me to squash all commits? Or you are going to do this anyway when you merge?

I'll do a squash merge.
Just waiting for our QA to give her stamp before proceeding.

alfetopito commented on libs/wallet/src/web3-react/hooks/useWalletMetadata.ts:
```diff
@@ -125,10 +124,7 @@ export function useIsSafeApp(): boolean {
-  const hasSafeInfo = !!useGnosisSafeInfo()
-
-  return !!safeAddress || hasSafeInfo
+  return !!useGnosisSafeInfo()
```
Actually, I am checking an issue with counterfactual Safes and found out that this was added for it 

https://github.com/cowprotocol/cowswap/pull/4063/files

Does your change account for counterfactual?
Meaning, does the SDK now return the info for counterfactual Safes?

alfetopito replied:
Well, it does work now 
![image](https://github.com/cowprotocol/cowswap/assets/43217/4d34be14-236a-491b-b9e1-841c90c7eeaf)

I guess this will fix Daniel's problem, then :)

compojoom replied:
I don't get it. It's not like this PR broke the counterfactual detection? The tests I did were with your current main release?

alfetopito replied:
This PR "fixed" counterfactual Safes (allowing them to be triggered) but at the same time "broke" the detection of them, given that they are now treated as EOAs.

compojoom replied:
Ok, I'll play a bit more to see if I can find some middle ground.

compojoom replied:
After debugging this with @anxolin and @alfetopito we discovered that the problem is in the useIsSmartContractWallet function and is independent of this PR:
https://github.com/cowprotocol/cowswap/blob/85a78bbfe9b0a8fe5428b25aba0b38944b10987d/libs/wallet/src/web3-react/hooks/useIsSmartContractWallet.ts#L26

With the changes in this PR though, one could call the useIsSafeWallet in the smartContractWallet and detect that this is a Safe(or a counterfactual safe). I think that the CoW team is also going to do some more detections there?

alfetopito replied:
Fix here https://github.com/cowprotocol/cowswap/pull/4553

elena-zh reviewed (approved):
Great job!

anxolin commented on libs/wallet/src/web3-react/updater.ts:
```diff
@@ -61,44 +66,83 @@ function _useWalletDetails(account?: string, standaloneMode?: boolean): WalletDe
-  }, [setSafeInfo, chainId, account, provider, isReadOnly])
+    return () => {
+      clearInterval(shortSafeInfoInterval)
+      clearInterval(longSafeInfoInterval);
```
should you nullify it so it doesn't happen that the promise above resolves after the hook is re-rendered?

anxolin commented on libs/wallet/src/web3-react/updater.ts:
```diff
@@ -61,44 +66,83 @@ function _useWalletDetails(account?: string, standaloneMode?: boolean): WalletDe
+              // Being here means that we don't run in an iframe over Safe's AppCommunicator,
+              // so we need to slow down the interval to not spam the RPC
+              clearInterval(shortSafeInfoInterval);
+              longSafeInfoInterval = setInterval(updateSafeInfo, SAFE_INFO_LONG_INTERVAL);
```
I wrote somewhere (maybe another PR?) that it would be nice to only create the intervals conditionally, right? 

I mean, if `chainId, account, provider` changes, the whole hook will be destroyed and re-rendered creating a new interval. Why would we want to execute at regular intervals the check if `chainId && account && provider` is false?

anxolin commented on apps/cowswap-frontend/src/modules/swap/pure/SwapButtons/SafeReadOnlyButton.tsx:
```diff
@@ -19,7 +19,7 @@ export function SafeReadOnlyButton() {
               Your Safe is not connected with a signer.
               <br />
-              To place an order, you must connect using a signer of the Safe and refresh the page.
+              To place an order, you must connect using a signer of the Safe.
```
👍

alfetopito reviewed (approved):
Code wise I'm happy with this.

Counterfactual Safes are not working as we discussed last week, but that's not coming from this PR, as it's already the case in prod.

I'll work on that on our side.

In the mean time, I asked our @elena-zh to check the Safe flows to make sure I did not miss anything.

alfetopito commented:
This should fix counterfactual Safe creation https://github.com/cowprotocol/cowswap/pull/4553

elena-zh commented:
Hey @compojoom , thanks for implementing this!
However, seems like it is not working as described: see, I'm not connected to a wallet in Safe, but the app does not show me 'Connect signer' message
![image](https://github.com/cowprotocol/cowswap/assets/70885163/702cb66f-b609-414f-bab7-72efde4b72e8)
Same happens when I'm not connected to a signer's account 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/fcb76735-6d9d-4266-8643-762b35043056)

Could you please double check your recent changes? 
Thanks!

elena-zh commented:
Also, I've noticed that bundling is not working in the current PR:
1. ETH-flow works like connected to EOA
![bundling](https://github.com/cowprotocol/cowswap/assets/70885163/42ca39dc-ff0b-4127-b8b1-bed02c6e5856)
2. Bundling (approve and swap) also behaves like connected to EOA:
![Screenshot_3](https://github.com/cowprotocol/cowswap/assets/70885163/6aed490b-c6da-4404-97e2-e1b716ad7911)
3. TWAP shows 'Unsupported wallet' 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/a4fc41ef-d7b5-4341-8fbc-423732502ab8)

compojoom commented:
Thanks for testing this @elena-zh . I see that my interval logic is wrong. I'll need to rethink this yet again :(

alfetopito commented on libs/wallet/src/web3-react/updater.ts:
```diff
@@ -65,40 +70,80 @@ function _useSafeInfo(walletInfo: WalletInfo): GnosisSafeInfo | undefined {
-          .then((_safeInfo) =>
-            setSafeInfo({
+        if (safeAppsSdk) {
+          const appsSdkSafeInfo = await safeAppsSdk.safe.getInfo()
+          setSafeInfo((prevSafeInfo) => {
+            const { isReadOnly, nonce } = appsSdkSafeInfo
```
As much as I prefer async/await instead of promise chaining, this is:
1. Not in the same style as the way it's fetched in the `else` clause
2. Not handling the failure case AFAICT

compojoom replied:
Yes, you are right. I just copied the old code that was using promises to the else block and added my code with await. I've changed this now, so both blocks are using async await syntax.

alfetopito reviewed (commented):
Thanks for quickly addressing the feedback.

I hope some of the issues Elena reported will be fixed by https://github.com/cowprotocol/cowswap/pull/4553. These 2 PRs are complementary to each other.

@elena-zh I've triggered the build again, should be ready to test the new changes.

elena-zh reviewed (approved):
Looks good to me now  (given the fact that all the rest issues are fixed in #4553)!

alfetopito commented:
@compojoom , I've merged my PR (https://github.com/cowprotocol/cowswap/pull/4553) to develop, that should hopefully fix the issues with counterfactual and Safe detection mentioned by Elena.

@elena-zh can you try again when you have time?

compojoom commented:
@elena-zh there is one edge case that I couldn't fix, but it is a problem in the web3-react/gnosis-safe package. There seems to be a race condition, but the user is only going to run into this if he in the safe with a counterfactual account on the "swap app" page and refreshes. Then the widget will say connect wallet :( But if you are in the safe in a counterfactual account and you navigate to the swap page it should be fine.

alfetopito commented:
I just synced your PR with develop where I fixed the lint issues, so I guess they come from this PR?

![image](https://github.com/cowprotocol/cowswap/assets/43217/7336a207-a12b-4ce9-8e37-618bac49d79c)

Other than that, everything is working (build, test)

compojoom commented:
Yes, that was me. Sorry. You guys need a pre-push like we do at safe-web: https://github.com/safe-global/safe-wallet-web/blob/dev/.husky/pre-push
It's very handy to prevent those lint problems :)

elena-zh commented:
Hey @compojoom , @alfetopito , changes related to the CoW Swap app LGTM!
However!
The issue I've reported in https://github.com/cowprotocol/cowswap/pull/4553#issuecomment-2174240158 (case 1) is still reproducible in the current PR. I've opened #4569 bug for this. 
And I have not been able to test these changes in the widget as I can see this error when I open preview link
![image](https://github.com/cowprotocol/cowswap/assets/70885163/a928f83a-d364-4bee-a9cf-70cbe98ee997)

compojoom commented:
@elena-zh maybe you need to test the dev version of app.safe.global as the latest release doesn't contain the changes we did to the iframe: https://safe-wallet-web.dev.5afe.dev/
Maybe with the latest dev safe and latest dev cow - it will work?

elena-zh commented:
Hey @compojoom , I've tested changes with:
1.[ PR preview ](https://swap-dev-git-fork-compojoom-fix-connect-signer-cowswap.vercel.app/#/11155111/swap/ETH/COW) connected to [dev.safe](https://safe-wallet-web.dev.5afe.dev/home?safe=sep%3A0xCdafC0C73D652FDE2ffeC7Bcb0f75e8fb4d14326) usibg WC, and I've got the same behavior as I reported on #4569 
2. https://dev.swap.cow.fi/ connected to the same safe above using WC: same result. 
3. I've also tested the case with dev.widget opened inside the dev safe, and it says 'connect a wallet' to me every time I open the app with no regard to a connected signer. Frankly, I thought that this issue happens only on my side (I reported it on https://github.com/cowprotocol/cowswap/issues/4229#issuecomment-2173441195 ).

alfetopito commented:
> Hey @compojoom , I've tested changes with: 1.[ PR preview ](https://swap-dev-git-fork-compojoom-fix-connect-signer-cowswap.vercel.app/#/11155111/swap/ETH/COW) connected to [dev.safe](https://safe-wallet-web.dev.5afe.dev/home?safe=sep%3A0xCdafC0C73D652FDE2ffeC7Bcb0f75e8fb4d14326) usibg WC, and I've got the same behavior as I reported on #4569 2. https://dev.swap.cow.fi/ connected to the same safe above using WC: same result. 3. I've also tested the case with dev.widget opened inside the dev safe, and it says 'connect a wallet' to me every time I open the app with no regard to a connected signer. Frankly, I thought that this issue happens only on my side (I reported it on [#4229 (comment)](https://github.com/cowprotocol/cowswap/issues/4229#issuecomment-2173441195) ).

The main question I have is: does this PR break the behaviour currently in prod?

If the answer is no and - although there's still more work to be done - it does improve the current situation, I'm in favor of merging it and addressing what's left in another PR.

elena-zh commented:
> > Hey @compojoom , I've tested changes with: 1.[ PR preview ](https://swap-dev-git-fork-compojoom-fix-connect-signer-cowswap.vercel.app/#/11155111/swap/ETH/COW) connected to [dev.safe](https://safe-wallet-web.dev.5afe.dev/home?safe=sep%3A0xCdafC0C73D652FDE2ffeC7Bcb0f75e8fb4d14326) usibg WC, and I've got the same behavior as I reported on #4569 2. https://dev.swap.cow.fi/ connected to the same safe above using WC: same result. 3. I've also tested the case with dev.widget opened inside the dev safe, and it says 'connect a wallet' to me every time I open the app with no regard to a connected signer. Frankly, I thought that this issue happens only on my side (I reported it on [#4229 (comment)](https://github.com/cowprotocol/cowswap/issues/4229#issuecomment-2173441195) ).
> 
> The main question I have is: does this PR break the behaviour currently in prod?
> 
> If the answer is no and - although there's still more work to be done - it does improve the current situation, I'm in favor of merging it and addressing what's left in another PR.

Agree. I've opened separate tasks to address issues we are discussing here as separate PRs.
